packages/rest-crud/src/crud-rest.controller.ts

```diff
@@ -0,0 +1,94 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/crud-rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {DataObject, Entity, EntityCrudRepository} from '@loopback/repository';
+import {getModelSchemaRef, post, requestBody} from '@loopback/rest';
+
+// Ideally, this file should simply `export class CrudRestController<...>{}`
+// Unfortunately, that's not possible for several reasons.
+//
+// First of all, to correctly decorate methods and define schemas for request
+// and response bodies, we need to know the target model which will be used by
+// the controller. As a result, this file has to export a function that will
+// create a constructor class specific to the given model.
+//
+// Secondly, TypeScript does not allow decorators to be used in class
+// expressions - see https://github.com/microsoft/TypeScript/issues/7342.
+// As a result, we cannot write implement the factory as `return class ...`,
+// but have to define the class as an internal type and return the controller
+// constructor in a new statement.
+// Because the controller class is an internal type scoped to the body of the
+// factory function, we cannot use it to describe the return type. We must
+// explicitly provide the return type.
+//
+// To work around those issues, we use the following design:
+// - The interface `CrudRestController` describes controller methods (members)
+// - The type `CrudRestControllerCtor` describes the class constructor.
+// - `defineCrudRestController` returns `CrudRestControllerCtor` type.
+
+export interface CrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T
+> {
+  readonly repository: EntityCrudRepository<T, IdType>;
+  create(_data: Omit<T, IdName>): Promise<T>;
+}
+
+export interface CrudRestControllerCtor<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T
+> {
+  new (repository: EntityCrudRepository<T, IdType>): CrudRestController<
+    T,
+    IdType,
+    IdName
+  >;
+}
+
+export function defineCrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T
+>(
+  modelCtor: typeof Entity & {prototype: T & {[key in IdName]: IdType}},
+): CrudRestControllerCtor<T, IdType, IdName> {
+  class CrudRestControllerImpl
+    implements CrudRestController<T, IdType, IdName> {
+    constructor(public readonly repository: EntityCrudRepository<T, IdType>) {}
+
+    @post('/', {
```

Comment:
Maybe expose `options` as an optional argument of `defineCrudRestController ` to customize paths and other things.

Reply:
Do we want to provide a `mixin` function to include CRUD controller methods?

Reply:
> Maybe expose `options` as an optional argument of `defineCrudRestController` to customize paths and other things.

Adding a new argument it's easy, let's leave that change for the future, if/when we actually need it.

> Do we want to provide a `mixin` function to include CRUD controller methods?

No. IMO, users should keep their controllers small. If they want to mix CRUD methods with some other APIs, then they should create two (or more) controllers.

Reply:
On the other thought, I see how `options` can be a better solution for customizing the `basePath` :+1: